#BRFSS cleaning example

The BRFSS is a fairly large and messy dataset. To put it into a useable format, you have to do a bit of cleaning and massaging. This Notebook demonstrates cleaning the data bit by bit for 2014. Scripts for every year are in the `cleaning_code` folder. I had to do a separate script for each year because the codebooks change slightly nearly every year, and it's just enough to make it a real headache.

Pandas is a terrific tool for data cleaning and manipulation, and that's what I use here.

##Getting started
Let's start by importing our libraries. We'll be using Pandas and Numpy.

In [1]:
import numpy as np
import pandas as pd

Next, we'll set our year variable (2014), which made it easy to change when I had to copy scripts for each year.

In [2]:
year = '2014'

##Reading in the data

Let's use Pandas to read in the `.csv` of the raw data for 2014. This isn't available on GitHub because the file is too big. Also, since the file is so big, you might get a warning.

Since we didn't specify UTF-8 encoding in the R script for creating the `.csv`'s, we need to specify the encoding, which somehow ends up being the interesting choice of "cp1252." That little tidbit right there was a couple hours worth of Googling.

In [4]:
df = pd.read_csv('data/brfss' + year + '.csv', encoding='cp1252')

/Users/wlarson/anaconda/lib/python3.4/site-packages/pandas/io/parsers.py:1130: DtypeWarning: Columns (121) have mixed types. Specify dtype option on import or set low_memory=False.
  data = self._reader.read(nrows)


Let's take a quick look at what we have here. As you can see, this is a rather large data set.

In [5]:
df

Unnamed: 0  x.state  fmonth     idate  imonth  iday  iyear  dispcode  \
0                1        1       1   1172014       1    17   2014      1100   
1                2        1       1   1072014       1     7   2014      1100   
2                3        1       1   1092014       1     9   2014      1100   
3                4        1       1   1072014       1     7   2014      1100   
4                5        1       1   1162014       1    16   2014      1100   
5                6        1       1   1022014       1     2   2014      1100   
6                7        1       1   1062014       1     6   2014      1100   
7                8        1       1   1112014       1    11   2014      1100   
8                9        1       1   1022014       1     2   2014      1100   
9               10        1       1   1082014       1     8   2014      1100   
10              11        1       1   1072014       1     7   2014      1100   
11              12        1       1   1022014       1     2   2014      1100   
12              13        1       1   1132014       1    13   2014      1100   
13              14        1       1   1042014       1     4   2014      1100   
14              15        1       1   1022014       1     2   2014      1100   
15              16        1       1   1062014       1     6   2014      1100   
16              17        1       1   1072014       1     7   2014      1100   
17              18        1       1   1072014       1     7   2014      1100   
18              19        1       1   1042014       1     4   2014      1100   
19              20        1       1   1082014       1     8   2014      1100   
20              21        1       1   1092014       1     9   2014      1200   
21              22        1       1   1032014       1     3   2014      1100   
22              23        1       1   1132014       1    13   2014      1100   
23              24        1       1   1152014       1    15   2014      1100   
24              25        1       1   1072014       1     7   2014      1100   
25              26        1       1   1142014       1    14   2014      1100   
26              27        1       1   1112014       1    11   2014      1100   
27              28        1       1   1082014       1     8   2014      1100   
28              29        1       1   1072014       1     7   2014      1100   
29              30        1       1   1142014       1    14   2014      1100   
...            ...      ...     ...       ...     ...   ...    ...       ...   
464634      464635       72      11  12102014      12    10   2014      1100   
464635      464636       72      11  12132014      12    13   2014      1100   
464636      464637       72      11  12112014      12    11   2014      1100   
464637      464638       72      11  12142014      12    14   2014      1100   
464638      464639       72      11  12102014      12    10   2014      1100   
464639      464640       72      11  12122014      12    12   2014      1100   
464640      464641       72      11  12122014      12    12   2014      1100   
464641      464642       72      11  12102014      12    10   2014      1100   
464642      464643       72      11  12142014      12    14   2014      1100   
464643      464644       72      11  12152014      12    15   2014      1200   
464644      464645       72      11  12132014      12    13   2014      1100   
464645      464646       72      11  12102014      12    10   2014      1100   
464646      464647       72      11  12122014      12    12   2014      1100   
464647      464648       72      11  12152014      12    15   2014      1100   
464648      464649       72      11  12152014      12    15   2014      1200   
464649      464650       72      11  12152014      12    15   2014      1100   
464650      464651       72      11  12102014      12    10   2014      1100   
464651      464652       72      11  12152014      12    15   2014      1100   
464652  

In [8]:
print("Number of rows:", len(df))
print("Number of columns (variables in the survey):", len(df.columns))
print(df.columns)

Number of rows: 464664
Number of columns (variables in the survey): 280
Index(['Unnamed: 0', 'x.state', 'fmonth', 'idate', 'imonth', 'iday', 'iyear', 'dispcode', 'seqno', 'x.psu', 'ctelenum', 'pvtresd1', 'colghous', 'stateres', 'ladult', 'numadult', 'nummen', 'numwomen', 'genhlth', 'physhlth', 'menthlth', 'poorhlth', 'hlthpln1', 'persdoc2', 'medcost', 'checkup1', 'exerany2', 'sleptim1', 'cvdinfr4', 'cvdcrhd4', 'cvdstrk3', 'asthma3', 'asthnow', 'chcscncr', 'chcocncr', 'chccopd1', 'havarth3', 'addepev2', 'chckidny', 'diabete3', 'diabage2', 'lastden3', 'rmvteth3', 'veteran3', 'marital', 'children', 'educa', 'employ1', 'income2', 'weight2', 'height3', 'numhhol2', 'numphon2', 'cpdemo1', 'internet', 'renthom1', 'sex', 'pregnant', 'qlactlm2', 'useequip', 'blind', 'decide', 'diffwalk', 'diffdres', 'diffalon', 'smoke100', 'smokday2', 'stopsmk2', 'lastsmk2', 'usenow3', 'alcday5', 'avedrnk2', 'drnk3ge5', 'maxdrnks', 'flushot6', 'flshtmy2', 'pneuvac3', 'shingle2', 'fall12mn', 'fallinj2', 'seatbelt

## Collecting our variables of interest

Great, so we have a lot of rows and 280 columns. We're just trying to find some basic demographic information. If we look through the codebook, we can find those variable names. Note that an underscore in the codebook is the same as `x.` in the data, and that variables are all caps in the codebook but lowercase in the data.

- **Income:** income2
- **Race:** x.race
- **State:** x.state
- **Age:** x.ageg5yr
- **Sex:** sex
- **Height:** height3
- **Weight:** weight2

We'll use height and weight to calculate BMI. In later years (like 2014), the good people at the CDC calculate that and include it as a separate variable, but in earlier years, they weren't so advanced, so we'll always just grab height and weight and calculate it ourselves.

We can use Pandas to easily grab the columns we want. We're going to grab those columns as Pandas Series, clean them up, and then piece everything back together as a master DataFrame.

In [9]:
income = df['income2']
race = df['x.race']
state = df['x.state']
age = df['x.ageg5yr']
sex = df['sex']
height = df['height3']
weight = df['weight2']

If we look at one of these variables, we can get a peek at the way the survey data is coded.

In [10]:
income

0      7
1      4
2      7
3      5
4      4
5      6
6      8
7      1
8     77
9      7
10    77
11     6
12     8
13     5
14     5
...
464649     8
464650     3
464651     1
464652     6
464653     5
464654     1
464655     2
464656    99
464657     2
464658     5
464659     6
464660     7
464661    99
464662     2
464663     4
Name: income2, Length: 464664, dtype: float64

##Cleaning up a column: Income

We'll start with cleaning up the income data. Looking at the codebook, the codes are:

* 1: <\$10k
* 2: \$10k - \$15k
* 3: \$15k - \$20k
* 4: \$20k - \$25k
* 5: \$25k - \$35k
* 6: \$35k - \$50k
* 7: \$50k - \$75k
* 8: \>\$75k
* 77: Unknown
* 99: Refused

We're going to group 77 and 99 together as `np.nan`, and we're going to group 7 and 8 together as simply \>\$50k (in earlier BRFSS years, it stops at />\$50K, rather than including the \$50k-\$75k and \>\$75k groups)

Pandas makes it easy to use a dictionary for replacement, so we'll put these codes into a dictionary.

In [11]:
income_replace = {1:'<10k', 2:'10k-15k', 3:'15k-20k', 4:'20k-25k', 5:'25k-35k', 6:'35k-50k', 7:'>50k', 8:'>50k', 77:np.nan, 99:np.nan}

##Cleaning the other columns

Next, we'll clean the race column. If you are working with historical data, this is the trickiest column. The name of the variable, the meanings of the codes, and the groups represented change frequently over the years, and sometimes in subtle ways. Being able to use historical data is why my dictionary ends up grouping some things together that the 2014 dataset breaks out. See the codebook for more details.

In [12]:
race_replace = {1:'white', 2:'black', 3:'native american', 4:'asian/pacific islander', 5:'asian/pacific islander', 6:'other/multiracial', 7:'other/multiracial', 8:'hispanic', 9:'refused/unknown'}

We'll skip the state column for now. It's represented as a FIPS statecode, and you can find more info on Wikipedia: [FIPS codes](https://en.wikipedia.org/wiki/Federal_Information_Processing_Standard_state_code).

Let's get the age and sex columns mapped.

In [13]:
age_replace = {1:'18-24', 2:'25-29', 3:'30-34', 4:'35-39', 5:'40-44', 6:'45-49', 7:'50-54', 8:'55-59', 9:'60-64', 10:'65-69', 11:'70-74', 12:'75-79', 13:'80+', 14:np.nan}
sex_replace = {1:'male', 2:'female'}

The height and weight columns are a little trickier to deal with, so we'll handle those in a minute.

##Using Pandas to do the replacement
Since we've got our dictionaries of codes mapped to values, we can let Pandas do the heavy lifting of cleaning up our columns.

In [14]:
income = income.replace(income_replace)
race = race.replace(race_replace)
age = age.replace(age_replace)
sex = sex.replace(sex_replace)

Let's check out one of our cleaned up columns. Now it is beautiful and easy-to-use:

In [15]:
income

0        >50k
1     20k-25k
2        >50k
3     25k-35k
4     20k-25k
5     35k-50k
6        >50k
7        <10k
8         NaN
9        >50k
10        NaN
11    35k-50k
12       >50k
13    25k-35k
14    25k-35k
...
464649       >50k
464650    15k-20k
464651       <10k
464652    35k-50k
464653    25k-35k
464654       <10k
464655    10k-15k
464656        NaN
464657    10k-15k
464658    25k-35k
464659    35k-50k
464660       >50k
464661        NaN
464662    10k-15k
464663    20k-25k
Name: income2, Length: 464664, dtype: object

##Cleaning the height and weight columns

Okay, now let's work with the height and weight data. We'll start by noting that 7777 and 9999 are 'unknown' or 'replaced' codes, so we'll NaN those right away. Also, in later years, the survey gave the option of taking height and weight in metric and imperial units. Earlier on, it's all imperial, so we'll make and option for using the metric system (which is denoted differently in the codes). Finally, our max weight changes over the years also, so we'll set that now to 999 lbs.

In [17]:
hw_replace = {7777:np.nan, 9999:np.nan}
metric = True
max_weight = 999
height = height.replace(hw_replace)
weight = weight.replace(hw_replace)

###Dealing with the height codes

Alright, now we can tackle processing the height column. In the codebook, we see that things are a little funny.

* 200-711: Height in ft/in (\_|\_ \_) - i.e. the first digit is feet, the second and third digits are inches. So 509 is 5' 9''
* 9000-9998: Height in m/cm (9\_|\_ \_)

To handle this, we are going to run through the height column, grab the value, and turn it into a string. We'll pop the first token and save it as the feet, save the other two tokens as the inches, and then we will cast everything as floats and convert the result to inches and then to meters.

If the number is greather than 9000 and our metric toggle is `True`, then we know that it is a metric value, and we'll do a similar process.

_**Note:** I noticed that if I do all of this in Pandas, it is very slow. It's much faster to put the Series into a python list, do the operations and append the result to a new list, and then save the result back to a Series._

In [18]:
height = height.tolist()
new_height = []
for row in height:
    h = str(row)

    if row < 1: # If the height is 0, it isn't real
        meters = np.nan

    elif row < 712:
        feet = float(h[0])
        inches = float(h[1:])
        if inches > 12: # Catches an error by the surveyor: no one is 5' 13" tall
            meters = np.nan
        else:
            inches = inches + feet*12
            meters = inches * 0.0254

    elif row < 9999 and row >= 9000 and metric:
        meters = float(h[1])+float(h[2:])*0.01
        if meters == 0: # If the height is 0, it isn't real
            meters = np.nan

    else: # If you've gotten this far, give up. You're NaN
        meters = np.nan

    new_height.append(meters)

Now let's do something similar for the weight. Except this time, we don't have to play fun games with strings for the imperial units (we still do with the metric units).

In [19]:
weight = weight.tolist()
new_weight = []
for row in weight:
    
    if row < 10: # No person ages 18-99 should weigh 10 lbs.
        kg = np.nan

    elif row < max_weight:
        kg = row * 0.453592

    elif row < 9999 and metric:
        w = str(row)
        kg = float(w[1:])
        if kg < 10: # Again, no person 18-99 should weigh 10 kgs. And yes, I know that 10 lbs != 10 kgs, but whatever
            kg = np.nan

    else:
        kg = np.nan

    new_weight.append(kg)

###Calculating BMI

We've got height and weight in metric units. Let's compute the BMI, which is 

$bmi=\frac{w}{h^2}$

BMIs that are less than 10 or greater than 200 are possible, but highly unlikely. At the moment, I'm not terribly interested in these outliers (which are honestly more likely the result of error than true BMIs anyway). We will exclude them.

In [20]:
bmi = []
for h, w in zip(new_height, new_weight):
    b = w/(h*h)
    if b < 10 or b > 200:
        b = np.nan
    bmi.append(b)

We'll quickly save the height, weight, and BMI back to Pandas Series.

In [21]:
height = pd.Series(new_height)
weight = pd.Series(new_weight)
bmi = pd.Series(bmi)

##Saving our clean data

We're almost there! Now, let's take all of our cleaned up columns and put them back together as a single DataFrame and name the columns something easy to understand. We'll save that out to a `.csv`.

In [22]:
brfss_out = pd.concat([income, race, state, age, sex, height, weight, bmi], axis=1)
brfss_out.columns = ['income', 'race', 'state', 'age', 'sex', 'height', 'weight', 'bmi']
brfss_out.to_csv('brfss' + year + 'clean.csv')

And now, at long last, we are finished. We have a cleaned up set of demographic data in `brfss2014clean.csv`. It is in the same format as what we can spit out for all of the other years of the BRFSS 1987-2014, so it will be easy to run analyses across all the years.